File: Java/HashMap/Find And Replace in String.java
class Solution {
    public String findReplaceString(String s, int[] indices, String[] sources, String[] targets) {
        HashMap<Integer, Pair> mp = new HashMap<>();
        int k = indices.length;
        for (int i = 0; i < k; i++) {
            int idx = indices[i];
            String src = sources[i];
            String targ = targets[i];

            if (idx + src.length() <= s.length() && s.substring(idx, idx + src.length()).equals(src)) {
                mp.put(idx, new Pair(src, targ));
            }
        }
        StringBuilder ans = new StringBuilder();
        int i = 0;
        while (i < s.length()) {
            if (mp.containsKey(i)) {
                Pair p = mp.get(i);
                ans.append(p.target); // append target
                i += p.source.length(); // skip source length
            } else {
                ans.append(s.charAt(i));
                i++;
            }
        }
        return ans.toString();
    }

    static class Pair {
        String source;
        String target;

        Pair(String s, String t) {
            source = s;
            target = t;
        }
    }
}

File: CPP/map/Find And Replace in String.cpp
class Solution {
public:
    string findReplaceString(string s, vector<int>& indices, vector<string>& sources, vector<string>& targets) {
        map<int, pair<string, string>> mp;
        int k = indices.size();
        for (int i = 0; i < k; i++) {
            int idx = indices[i];
            string src = sources[i];
            string targ = targets[i];

            if (s.substr(idx, src.size()) == src) {
                mp[idx] = {src, targ};
            }
        }

        string ans;
        int i = 0;
        while (i < s.size()) {
            // If there is a replacement starting at position i
            if (mp.find(i) != mp.end()) {
                ans += mp[i].second;  // append target string
                i += mp[i].first.size();  // skip the source length
            } else {
                ans += s[i];  // no replacement, just copy current char
                i++;
            }
        }
        
        return ans;

    }
};
